hdat/reports.py

```diff
@@ -0,0 +1,57 @@
+import csv
+import sys
+
+from collections import defaultdict
+from itertools import chain
+
+
+def expand_key_parts(result, nested_key):
+    if nested_key[-1] == "*":
+        nested_data = get_nested_item(result, nested_key[:-1])
+        if isinstance(nested_data, dict):
+            for key in nested_data:
+                yield nested_key[:-1] + [key]
+    else:
+        yield nested_key
+
+
+def get_nested_item(result, nested_key):
+    num_levels = len(nested_key)
+    try:
+        if num_levels == 0 or nested_key[0] not in result:
+            return
+        elif num_levels == 1:
+            return result[nested_key[0]]
+        else:
+            return get_nested_item(result[nested_key[0]], nested_key[1:])
+    except TypeError:
+        error_msg = 'Result value "{}" is not iterable and has no specifier "{}"\n'
+        sys.stderr.write(error_msg.format(result, nested_key[0]))
+        return
+
+
+def print_results(results, input_keys_str):
+    if not input_keys_str:
+        input_keys_str = 'case_id,result_id,ran_on,commit,metrics.*'
+    reader_keys = csv.reader([input_keys_str])

```

Comment:
This is very clever---I hadn't considered the possibility of there being commas in the keys themselves (only the values).

Perhaps, since we are using the csv read, we no longer need to call `key.strip()` on each value?

Reply:
I tried it without `key.strip()` and it kept spaces after and before the keys, so I think it might still be necessary unless the user knows to input keys in csv format. Otherwise, if they enter `--keys "test, test2, test3"`, it will fail to strip the space in front.
![image](https://user-images.githubusercontent.com/9055029/43207447-c2370634-8fed-11e8-835b-012ef21dd98c.png)

Top with `key.strip`, bottom without.

Reply:
I see, then lets keep the `key.strip()`.  It seems like it will save headaches for users who accidentally add spaces (although it will prevent users from adding a space if they actually want one---but I can't imagine why anyone would want a space in from of their keys).